spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/MappingSpannerReadConverter.java

```diff
@@ -73,163 +78,201 @@
+	private void attemptToReadProperty(StructAccessor structAccessor,
+			SpannerPersistentProperty spannerPersistentProperty,
+			PersistentPropertyAccessor accessor) {
+		String columnName = spannerPersistentProperty.getColumnName();
+		Object value = structAccessor.get(spannerPersistentProperty);
+		if (value != null) {
+			accessor.setProperty(spannerPersistentProperty, value);
+		}
+		else {
+			throw new SpannerDataException(String.format(
+					"The value in column with name %s"
+							+ " could not be converted to the corresponding property in the entity."
+							+ " The property's type is %s.",
+					columnName, spannerPersistentProperty.getType()));
+		}
 	}
 
-	private boolean attemptReadSingleItemValue(
-			SpannerPersistentProperty spannerPersistentProperty, Struct struct,
-			Class sourceType,
-			String colName, PersistentPropertyAccessor accessor) {
-		Class targetType = spannerPersistentProperty.getType();
-		if (sourceType == null || !canConvert(sourceType, targetType)) {
-			return false;
-		}
-		BiFunction readFunction = singleItemReadMethodMapping
-				.get(ConversionUtils.boxIfNeeded(sourceType));
-		if (readFunction == null) {
-			return false;
+	private <R> boolean shouldSkipProperty(StructAccessor structAccessor,
+			SpannerPersistentProperty spannerPersistentProperty, Set<String> includeColumns,
+			boolean readAllColumns,
+			PreferredConstructor<? extends R, SpannerPersistentProperty> persistenceConstructor,
+			boolean allowMissingColumns) {
+		String columnName = spannerPersistentProperty.getColumnName();
+		boolean notRequiredByPartialRead = !readAllColumns && !includeColumns.contains(columnName);
+		boolean isConstructorParameter = persistenceConstructor != null
+				&& persistenceConstructor.isConstructorParameter(spannerPersistentProperty);
+
+		return notRequiredByPartialRead
+				|| isValidMissingColumn(structAccessor, allowMissingColumns, spannerPersistentProperty.getColumnName())
+				|| structAccessor.isNull(columnName)
+				|| isConstructorParameter;
+	}
+
+	private boolean isValidMissingColumn(StructAccessor structAccessor, boolean allowMissingColumns,
+			String columnName) {
+		if (!structAccessor.hasColumn(columnName)) {
+			if (!allowMissingColumns) {
+				throw new SpannerDataException(
+						"Unable to read column from Spanner results: "
+								+ columnName);
+			}
+			else {
+				return true;
+			}
 		}
-		accessor.setProperty(spannerPersistentProperty,
-				convert(readFunction.apply(struct, colName), targetType));
-		return true;
+		return false;
+	}
+
+	private <R> PersistentEntityParameterValueProvider<SpannerPersistentProperty> getParameterValueProvider(
+			SpannerPersistentEntity<? extends R> persistentEntity, StructAccessor structAccessor) {
+		return new PersistentEntityParameterValueProvider<>(
+				persistentEntity, new PropertyValueProvider<SpannerPersistentProperty>() {
+					@Override
+					public <T> T getPropertyValue(SpannerPersistentProperty property) {
```

Comment:
the type param T is only used to cast the return and is not actually restricting anything because none of the arguments or methods called inside this method restrict T. You could just return Object type from getPropertyValue